File: sdks/py/src/loxa/core/params.py
from __future__ import annotations

from dataclasses import replace

from .event import Params


def http_params(method: str, path: str, *, service: str = "", route: str = "", name: str = "") -> Params:
    return Params(
        service=service,
        name=name or f"{method} {path}",
        event=name or f"{method} {path}",
        kind="http",
        method=method,
        path=path,
        route=route,
    )


def job_params(name: str, *, service: str = "") -> Params:
    return Params(service=service, name=name, event=name, kind="job")


def queue_params(queue: str, message_id: str = "", *, service: str = "") -> Params:
    return Params(service=service, name=queue, event=queue, kind="queue", request_id=message_id)


def cli_params(command: str, *, service: str = "") -> Params:
    return Params(service=service, name=command, event=command, kind="cli")


def cron_params(job: str, *, service: str = "") -> Params:
    return Params(service=service, name=job, event=job, kind="cron")


def with_trace(params: Params, trace_id: str, span_id: str = "") -> Params:
    return replace(params, trace_id=trace_id, span_id=span_id)


__all__ = [
    "Params",
    "http_params",
    "job_params",
    "queue_params",
    "cli_params",
    "cron_params",
    "with_trace",
]

File: sdks/go/src/core/params.go
package core

import "time"

// Params holds all inputs for starting a canonical event.
// Canonical fields are first-class struct members; extra business context
// goes in Custom or is added later via Enrich.
type Params struct {
	// ── Identity ────────────────────────────────────────────────────────────
	Event   string
	Name    string
	Kind    string
	Message string
	Level   Level

	// ── Correlation IDs ──────────────────────────────────────────────────────
	RequestID string
	TraceID   string
	SpanID    string
	ParentID  string

	// ── Service metadata ─────────────────────────────────────────────────────
	Service      string
	Version      string
	Environment  string
	DeploymentID string
	Region       string
	Host         string
	Runtime      string

	// ── Request metadata ─────────────────────────────────────────────────────
	Method     string
	Path       string
	Route      string
	StatusCode int
	DurationMS int64
	Outcome    string

	// ── Canonical subject identifiers ───────────────────────────────────────
	UserID         string
	TenantID       string
	WorkspaceID    string
	OrganizationID string
	SessionID      string

	// ── Timing ───────────────────────────────────────────────────────────────
	StartedAt time.Time

	// ── Custom business context ───────────────────────────────────────────────
	// Custom holds domain-specific attrs that LOXA-Go does not know in advance.
	// These are copied into Event.Attrs during StartEvent.
	Custom []Attr
}

// With returns a copy of p with additional attrs appended to Custom.
// It enables a builder style:
//
//	loxa.Params{Event: "checkout.request"}.With(
//	    loxa.String("tenant.id", tenantID),
//	)
func (p Params) With(attrs ...Attr) Params {
	p.Custom = append(p.Custom, attrs...)
	return p
}
